### ISSUE TITLE
com.orientechnologies.common.exception.OException: File flush was abnormally terminated

### ISSUE THREAD

**Issue by thiagopm** [orientechnologies/orientdb#4610]
Hello,

We created a function to save a list of an Edge where we send a RID (FeedRID) and a list of users related with this. (5.000 users/request)

After some requests, this error below started. Then We should to restore a new database.

Do you know why it could be happened?

Error:
<code>
com.orientechnologies.common.exception.OException: File flush was abnormally terminated
java.util.concurrent.ExecutionException: java.io.IOException: Stream Closed
java.io.IOException: Stream Closed
</code>

function:
<code>
if(feedRID == null)</br>
{</br>
	response.send(404, "Invalid feedRID", "text/plain", "Error: Invalid feedRID" );</br>
}</br>
else</br>
{</br>
	var db = orient.getGraph();</br>
	</br>
    try{</br>
      	var feed = db.getVertex(feedRID);</br>
      </br>
      	var userArray = userList.split(",");</br>
      </br>
      	var index;</br>
      	for (index = 0; index < userArray.length; ++index) {</br>
      		var userId = userArray[index];</br>
          	</br>
          var d = new Date,
              dformat = [d.getFullYear(),
               d.getMonth()+1,
               d.getDate()].join('-')+' '+
              [d.getHours(),
               d.getMinutes(),
               d.getSeconds()].join(':');</br>
          </br>
          var feedUser = db.addVertex("class:FeedUser");</br>
          feedUser.setProperty('FeedId', 0)</br>
          feedUser.setProperty('FeedUserId', 0)</br>
          feedUser.setProperty('LastActionDate', dformat)</br>
          feedUser.setProperty('UserId', userId)</br>
                              </br>
        	db.addEdge(null, feed, feedUser, "FeedUserFeed");</br>
      	}</br>
      
      	return "Ok1";</br>
    }</br>
  	catch ( err ){</br>
    	response.send(500, "Error on creating FeedUser", "text/plain", err.toString() );</br>
    }</br>
}</br>
</code>

Thank You,
Thiago

**Comment by valeriotarenghi** [comment]
Hi Thiago
Can you send me your function?

Thanks
Valerio

**Comment by thiagopm** [comment]
of course:

function SaveFeedUser(feedRID, userList)
{

if(feedRID == null)
{
	response.send(404, "Invalid feedRID", "text/plain", "Error: Invalid feedRID" );
}
else
{
	var db = orient.getGraph();
	
    try{
      	var feed = db.getVertex(feedRID);
      
      	var userArray = userList.split(",");
      
      	var index;
      	for (index = 0; index < userArray.length; ++index) {
      		var userId = userArray[index];
          	
          var d = new Date,
              dformat = [d.getFullYear(),
               d.getMonth()+1,
               d.getDate()].join('-')+' '+
              [d.getHours(),
               d.getMinutes(),
               d.getSeconds()].join(':');          
          
          var feedUser = db.addVertex("class:FeedUser");
          feedUser.setProperty('FeedId', 0)
          feedUser.setProperty('FeedUserId', 0)
          feedUser.setProperty('LastActionDate', dformat)
          feedUser.setProperty('UserId', userId)
                               
        	db.addEdge(null, feed, feedUser, "FeedUserFeed");
      	}
      
      	return "Ok1";
    }
  	catch ( err ){
    	//db.rollback();
		response.send(500, "Error on creating FeedUser", "text/plain", err.toString() );
    }
}

}

**Comment by valeriotarenghi** [comment]
Hi Thiago could you send me your schema structure for replicate your problem?

**Comment by thiagopm** [comment]
Ok, I'm sending only classes used in the function, if you need the complete schema ask me please.

CREATE CLASS User EXTENDS V
CREATE PROPERTY User.UserId INTEGER
CREATE PROPERTY User.UserLoginId INTEGER
CREATE PROPERTY User.UserOrganizationId INTEGER
CREATE PROPERTY User.OrganizationId INTEGER
CREATE PROPERTY User.IsAdmin BOOLEAN

CREATE CLASS Feed EXTENDS V
CREATE PROPERTY Feed.FeedId INTEGER
CREATE PROPERTY Feed.Title STRING
CREATE PROPERTY Feed.Description STRING
CREATE PROPERTY Feed.FeedTypeId INTEGER
CREATE PROPERTY Feed.TaskTypeId INTEGER
CREATE PROPERTY Feed.CreateDate DATETIME
CREATE PROPERTY Feed.LessonCategoryId INTEGER
CREATE PROPERTY Feed.URL STRING
CREATE PROPERTY Feed.CardHtml STRING
CREATE PROPERTY Feed.UserId INTEGER
CREATE PROPERTY Feed.UserName STRING
CREATE PROPERTY Feed.ParentId INTEGER

CREATE CLASS FeedUser EXTENDS V 
CREATE PROPERTY FeedUser.FeedUserId INTEGER
CREATE PROPERTY FeedUser.FeedId INTEGER
CREATE PROPERTY FeedUser.UserId INTEGER
CREATE PROPERTY FeedUser.LastActionDate DATETIME

CREATE CLASS FeedUserFeed EXTENDS E

CREATE CLASS UserFeedUser EXTENDS E

Thanks

**Comment by thiagopm** [comment]
Ow...I forgot this indexes:
CREATE INDEX User.UserId NOTUNIQUE_HASH_INDEX
CREATE INDEX Feed.FeedId NOTUNIQUE_HASH_INDEX
CREATE INDEX FeedUser.UserId NOTUNIQUE_HASH_INDEX

**Comment by LucaSon90** [comment]
Hi Thiago, could you tell me which version are you using ?

**Comment by LucaSon90** [comment]
We tested your function with your schema.
Try to do this changes:
1) adding db.commit () after you add the edge in the function;
2) change the type of property UserId in the FeedUser class from Integer to String;

Passing these parameters:
→ the RID of the feed (e.g. 13: 0)
→ the list of user (like this 12:0,12:1,12: 2)
the function generates the required edge.

**Comment by laa** [comment]
Hi, this is bug which we chaining for a long time. I will provide distribution to find reason of it.

**Comment by laa** [comment]
Please do not change code till we find root reason of issue, in any case, does not matter which code you use db should not be broken.

**Comment by laa** [comment]
Could you do following:

1. Use this distrtibution https://www.dropbox.com/s/xs2w59sncv9ln4m/orientdb-community-2.1-SNAPSHOT-distribution.zip?dl=0 
2. Set system property during server JVM start -Dfile.trackFileClose .
3. Repeat your test till you get java.io.IOException: Stream Closed 
4. Send me a.lomakin@orientdb.com server log.

I will look on it fix your issue.

**Comment by thiagopm** [comment]
@LucaSon90, I'm using v 2.1 rc5. Thank you.

@laa, I'll do it... Any news I tell you.

Thanks.

**Comment by laa** [comment]
Hi,
Any updates on this ?

**Comment by laa** [comment]
It is very critical issue for us, so once I have your data I will start to fix it.

**Comment by thiagopm** [comment]
Hi Andrey,
I sent you an email with a dump and logs, did you receive?
We got another error, but not this "stream closed".
We are trying, any update I tell you

**Comment by laa** [comment]
Hi ,
Sorry but why do you think that OrientDB is blocked ?
I see that all threads waiting for schema reload, but this thread (reload of schema) is not frozen. It is working.
Also I see that you have huge amount of threads , it does not improve performance, it introduces huge context switching overhead.

**Comment by laa** [comment]
I will provide you new distribution in few minutes, but still, are you sure that this is OK for you to use such amount of threads ?

**Comment by laa** [comment]
Could you try this one https://www.dropbox.com/s/xs2w59sncv9ln4m/orientdb-community-2.1-SNAPSHOT-distribution.zip?dl=0  it contains fix to resolve you current bottleneck.
But still that is super important to reproduce stream closed exception.

**Comment by thiagopm** [comment]
OrientDB wasn't blocked, but wasn't answering the requests for almos an hour....

I can reduce pool size, today it's 40. 
As we have 2 App servers and depending on the traffic we can have 20 Web servers.
Number of threads can go to 800.
Furthermore, the .Net driver always creates more connections when there isn't a free in the pool.

**Comment by thiagopm** [comment]
Ok, I'll start this version in some minutes.

**Comment by laa** [comment]
I will be partially online so keep me posted.

**Comment by laa** [comment]
This issue is only left for release so all development team is looking forward to fix this issue :-)

**Comment by thiagopm** [comment]
Man, We are trying to get this error again, but for while no way... 
I'll keep it on till tomorrow... Sometimes we're getting only some http 500 when calling the function..

**Comment by lvca** [comment]
If each function creates 5k edges (5.000 users/request as you mentioned), could it be an out of memory problem? How much RAM do you have? How many concurrent call can be done to this function? Could you split the for() cycle to commit every X items, where X could be 500 for now? Example:

```js
try{
    var feed = db.getVertex(feedRID);

    var userArray = userList.split(",");

    var index;
    for (index = 0; index < userArray.length; ++index) {
        var userId = userArray[index];

      var d = new Date,
          dformat = [d.getFullYear(),
           d.getMonth()+1,
           d.getDate()].join('-')+' '+
          [d.getHours(),
           d.getMinutes(),
           d.getSeconds()].join(':');          

      var feedUser = db.addVertex("class:FeedUser");
      feedUser.setProperty('FeedId', 0)
      feedUser.setProperty('FeedUserId', 0)
      feedUser.setProperty('LastActionDate', dformat)
      feedUser.setProperty('UserId', userId)

        db.addEdge(null, feed, feedUser, "FeedUserFeed");

      if( index % 500 == 0 ) {
        db.commit();
        db.begin();
       } 
    }
    db.commit();

    return "Ok1";
}
catch ( err ){
    //db.rollback();
    response.send(500, "Error on creating FeedUser", "text/plain", err.toString() );
}
```

**Comment by thiagopm** [comment]
Hi @lvca,

Yes we are having some memory problems either because of this function or some new queries. todays server have 15GB and we were getting this message since yesterday:

2015-07-23 07:13:05:506 INFO  -> Open server.sh (or server.bat on Windows) and c
hange the following variables: 1) MAXHEAP=-Xmx1434M 2) MAXDISKCACHE=11876 [OAbst
ractProfiler$MemoryChecker]

Today we got this:

2015-07-23 09:43:43:395 WARNI {db=ilang} Record (null) is null [OrientEdgeIterat
or]java.lang.OutOfMemoryError: GC overhead limit exceeded
Dumping heap to java_pid336.hprof ...
Heap dump file created [3224542164 bytes in 24.264 secs]

I talked to our developer that is developing this feature about some improvements on this, but as we were trying to get "java.io.IOException: Stream Closed" error, we hadn't changed yet. 

Thanks of the tips.
